app/src/org/commcare/android/tasks/TemplatePrinterTask.java

```diff
@@ -0,0 +1,253 @@
+
+
+    public TemplatePrinterTask(File input, String outputPath, Bundle values,
+                               PopulateListener listener) {
+        this.inputFile = input;
+        this.outputPath = outputPath;
+        this.values = values;
+        this.listener = listener;
+    }
+
+    /**
+     * Attempts to perform population of the template file, and throws the appropriate exception
+     * if encountering an error.
+     */
+    @Override
+    protected Integer doInBackground(Void... params) {
+        try {
+            populateHtml(inputFile, values);
+            return SUCCESS;
+        } catch (IOException e) {
+            return IO_ERROR;
+        } catch (RuntimeException e) {
+            problemString = e.getMessage();
+            return validationErrorType;
+        }
+    }
+
+    /**
+     * Receives the return value from doInBackground and proceeds accordingly
+     */
+    @Override
+    protected void onPostExecute(Integer result) {
+        listener.onFinished(result, problemString);
+    }
+
+    /**
+     * Populates an html print template based on the given set of key-value pairings
+     * and save the newly-populated template to a temp location
+     *
+     * @param input the html print template
+     * @param values the mapping of keywords to case property values
+     * @throws IOException
+     */
+    private void populateHtml(File input, Bundle values) throws IOException {
+        // Read from input file
+        // throws IOException
+        String fileText = TemplatePrinterUtils.docToString(input);
+
+        // Check if <body></body> section of html string is properly formed
+        // throws RuntimeException
+        int startBodyIndex = fileText.indexOf("<body");
+        String beforeBodySection = fileText.substring(0, startBodyIndex);
+        String bodySection = fileText.substring(startBodyIndex);
+        validateStringOrThrowException(bodySection);
+
+        // Swap out place-holder keywords for case property values within <body></body> section
+        bodySection = replace(bodySection, values);
+
+        // Write the new HTML to the desired  temp file location
+        // throws IO Exception
+        TemplatePrinterUtils.writeStringToFile(beforeBodySection + bodySection, outputPath);
+    }
+
+    /**
+     * Populate an input string with attribute keys formatted as {{ attr_key }}
+     * with attribute values.
+     * @param input String input
+     * @param values Bundle of String attribute key-value mappings
+     * @return The populated String
+     */
+    private static String replace(String input, Bundle values) {
+        // Split input into tokens bounded by {{ and }}
+        String[] tokens = TemplatePrinterUtils.splitKeepDelimiter(input, "\\{{2}", "\\}{2}");
+        for (int i=0; i<tokens.length; i++) {
+            String token = tokens[i];
+
+            // Every 2nd token is a attribute enclosed in {{ }}
+            if (i % 2 == 1) {
```

Comment:
Quoting this warning from the FindBugs Android Studio plugin:
The code uses x % 2 == 1 to check to see if a value is odd, but this won't work for negative numbers (e.g., (-5) % 2 == -1). If this code is intending to check for oddness, consider using x & 1 == 1, or x % 2 != 0.